### PR TITLE
Various fixes and removed deprecations

### PR DESCRIPTION
**Pull request by alexandre-daubois**
First batch of fixes with the "new" CI.

### CODE REVIEW

File: tests/Mbstring/MbstringTest.php [hunk 1]
```diff
@@ -68,13 +68,18 @@ public function testInternalEncodingWithInvalidEncoding()
...
-        $this->assertSame('&#23455;<&>d&eacute;j&agrave;', mb_convert_encoding('実<&>déjà', 'Html-entities'));
-        $this->assertSame('déjà', mb_convert_encoding(base64_encode('déjà'), 'Utf-8', 'Base64'));
-        $this->assertSame('déjà', mb_convert_encoding('d&eacute;j&#224;', 'Utf-8', 'Html-entities'));
-        $this->assertSame('déjà', mb_convert_encoding(utf8_decode('déjà'), 'Utf-8', 'ASCII,ISO-2022-JP,UTF-8,ISO-8859-1'));
-        $this->assertSame('déjà', mb_convert_encoding(utf8_decode('déjà'), 'Utf-8', ['ASCII', 'ISO-2022-JP', 'UTF-8', 'ISO-8859-1']));
+        $this->assertSame(mb_convert_encoding('déjà', 'ISO-8859-1', 'UTF-8'), mb_convert_encoding('déjà', 'Windows-1252'));
+        $this->assertSame('déjà', mb_convert_encoding(mb_convert_encoding('déjà', 'ISO-8859-1', 'UTF-8'), 'Utf-8', 'ASCII,ISO-2022-JP,UTF-8,ISO-8859-1'));
+        $this->assertSame('déjà', mb_convert_encoding(mb_convert_encoding('déjà', 'ISO-8859-1', 'UTF-8'), 'Utf-8', ['ASCII', 'ISO-2022-JP', 'UTF-8', 'ISO-8859-1']));
+
+        if (\PHP_VERSION_ID < 80200) {
```
Comment (nicolas-grekas) [inline tests/Mbstring/MbstringTest.php]: maybe putting this in a dedicated `@group legacy` would work?

**Comment by derrabus** [comment]
> There's one `utf8_decode()` occurrence left in Iconv. The recommended replacement `mb_convert_encoding($string, 'ISO-8859-1', 'UTF-8')` doesn't work in this case, so it deserves a bit more research. However, this fixes some problems here and there already.

We used to have a polyfill for `utf8_decode()` in the now-removed PHP 7.2 polyfill package. Would it help if we copied the code of that polyfill over to the iconv package?

File: tests/Mbstring/MbstringTest.php (hunk 1 above)
Comment (alexandre-daubois) [reply tests/Mbstring/MbstringTest.php]: It does. Thanks!

**Comment by alexandre-daubois** [comment]
Indeed @derrabus, that's smart. I copied the function and added it to Iconv as a private method. Seems to work really well 👍

File: src/Intl/Idn/bootstrap80.php [hunk 2]
```diff
@@ -118,8 +118,8 @@
 }
 
 if (!function_exists('idn_to_ascii')) {
-    function idn_to_ascii(?string $domain, ?int $flags = IDNA_DEFAULT, ?int $variant = INTL_IDNA_VARIANT_UTS46, &$idna_info = null): string|false { return p\Idn::idn_to_ascii((string) $domain, (int) $flags, (int) $variant, $idna_info); }
+    function idn_to_ascii(string $domain, int $flags = IDNA_DEFAULT, int $variant = INTL_IDNA_VARIANT_UTS46, &$idna_info = null): string|false { return p\Idn::idn_to_ascii($domain, $flags, $variant, $idna_info); }
 }
 if (!function_exists('idn_to_utf8')) {
-    function idn_to_utf8(?string $domain, ?int $flags = IDNA_DEFAULT, ?int $variant = INTL_IDNA_VARIANT_UTS46, &$idna_info = null): string|false { return p\Idn::idn_to_utf8((string) $domain, (int) $flags, (int) $variant, $idna_info); }
+    function idn_to_utf8(string $domain, int $flags = IDNA_DEFAULT, int $variant = INTL_IDNA_VARIANT_UTS46, &$idna_info = null): string|false { return p\Idn::idn_to_utf8($domain, $flags, $variant, $idna_info); }
```
Comment (stof) [inline src/Intl/Idn/bootstrap80.php]: Those changes should be reverted, as it makes it incompatible with passing `null` in non-strict-types files (which is deprecated in PHP 8.1+ but will only be removed in PHP 9.0)

File: src/Iconv/Iconv.php [hunk 3]
```diff
@@ -741,4 +741,34 @@ private static function getData($file)
 
         return false;
     }
+
+    private static function utf8Decode($s)
```
Comment (stof) [inline src/Iconv/Iconv.php]: I suggest naming `utf8ToLatin1` instead, to reflect what this method actually does (the bad name of `utf8_decode` was the main reason why it has been deprecated)

Comment (stof) [reply src/Iconv/Iconv.php]: Alternatively, we could drop `strlen1` entirely, falling back directly to the `strlen2` implementation that was used when the `xml` extension was not loaded and so we did not have the native `utf8_decode` function. The `utf8_decode` fallback was done because using the native function was faster than the userland logic of `strlen2`, but this probably does not hold anymore with a userland function doing more work.

File: tests/Php81/Php81Test.php [hunk 4]
```diff
@@ -34,6 +34,7 @@ public function testArrayIsList()
 
     /**
      * @requires extension mysqli
+     * @requires PHP < 8.4
```
Comment (derrabus) [inline tests/Php81/Php81Test.php]: This test should be executed on PHP 8.4 as well. We should rather allow the test to trigger deprecations.

Comment (alexandre-daubois) [reply tests/Php81/Php81Test.php]: Right

File: src/Intl/Idn/bootstrap80.php (hunk 2 above)
Comment (alexandre-daubois) [reply src/Intl/Idn/bootstrap80.php]: I reverted the casts

File: src/Iconv/Iconv.php (hunk 3 above)
Comment (alexandre-daubois) [reply src/Iconv/Iconv.php]: I went for the alternative solution, which really clean things up.

File: tests/Php82/SensitiveParameterValuePolyfillTest.php [hunk 5]
```diff
@@ -17,7 +17,7 @@
 /**
  * @author Tim Düsterhus <duesterhus@woltlab.com>
  */
-class SensitiveParameterValueTest extends TestCase
+class SensitiveParameterValuePolyfillTest extends TestCase
```
Comment (alexandre-daubois) [inline tests/Php82/SensitiveParameterValuePolyfillTest.php:20]: TestListenerTrait tries to find a polyfill declaration in `bootstrap.php` base on the test classname. However, `SensitiveParameterValue` is a "special" polyfill that doesn't need any declaration in `bootstrap.php`.

So, PHPUnit warned that no polyfill was found, but that's normal. By renaming, the file to not match the `SensitiveParametrerValue` classname, we avoid `TestListenerTrait` to raise a wrong warning.

File: src/Intl/Idn/bootstrap80.php (hunk 2 above)
Comment (stof) [reply src/Intl/Idn/bootstrap80.php]: You need to revert the whole change. The most important part of my comment is about the parameter type.
PHP does not support passing `null` to a non-nullable *userland* function parameter. The reason of the PHP 8.1 deprecation is precisely because the behavior of native functions cannot be emulated exactly in userland for now (in PHP 9, they will behave the same than userland functions in that regard)

Comment (stof) [reply src/Intl/Idn/bootstrap80.php]: @nicolas-grekas what do you think about adding legacy tests covering passing `null` to all our polyfilled functions (that will have a non-nullable parameter type in PHP 9), so that the testsuite catches such broken changes to the polyfills, as that's not the first time we get such changes in a PR.

Comment (alexandre-daubois) [reply src/Intl/Idn/bootstrap80.php]: I see what you mean, but what about [this warning](https://github.com/symfony/polyfill/actions/runs/10795014872/job/29940652441#step:5:733)? This is the reason I updated the signature

Comment (stof) [reply src/Intl/Idn/bootstrap80.php]: @alexandre-daubois this should have been fixed by https://github.com/symfony/polyfill/pull/495 AFAIK

Comment (alexandre-daubois) [reply src/Intl/Idn/bootstrap80.php]: It doesn't seem to be touching anything related to the `Incompatible signature for PHP >= 8` warning. Am I missing something?

Comment (stof) [reply src/Intl/Idn/bootstrap80.php]: it touches the detection logic at https://github.com/symfony/polyfill/pull/495/files#diff-05ed2f919a0a2d68912601230246f2c39077cd90c3ac9df8ac74651dd7c197aeR140

Comment (alexandre-daubois) [reply src/Intl/Idn/bootstrap80.php]: Oh right I get it now. That wasn't the root cause of the warning message. That's reverted 👍

File: tests/Mbstring/MbstringTest.php [hunk 6]
```diff
@@ -68,13 +68,21 @@ public function testInternalEncodingWithInvalidEncoding()
      */
     public function testConvertEncoding()
     {
-        $this->assertSame(utf8_decode('déjà'), mb_convert_encoding('déjà', 'Windows-1252'));
+        $this->assertSame(mb_convert_encoding('déjà', 'ISO-8859-1', 'UTF-8'), mb_convert_encoding('déjà', 'Windows-1252'));
```
Comment (xabbuh) [inline tests/Mbstring/MbstringTest.php]: Should we really replace `utf8_decode()` with `mb_convert_encoding()` here as `mb_convert_encoding()` is the polyfill that we test here, right?

Comment (alexandre-daubois) [reply tests/Mbstring/MbstringTest.php]: That's right. I moved the test case in a separate method to still use `utf8_decode()`, marked as `@group legacy`.

Comment (stof) [reply tests/Mbstring/MbstringTest.php]: Or use `iconv`

Comment (alexandre-daubois) [reply tests/Mbstring/MbstringTest.php]: Suits me as well, PR updated

**Review by xabbuh — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @alexandre-daubois.
